linkerd/proxy/http/src/client_handle.rs

```diff
@@ -0,0 +1,97 @@
+use std::{
+    net::SocketAddr,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::sync::Notify;
+
+/// A server-set extension that holds information about the client.
+#[derive(Clone, Debug)]
+pub struct ClientHandle {
+    /// The peer address of the client.
+    pub addr: SocketAddr,
+
+    /// Notifies the client to shutdown its connection.
+    pub close: Close,
+}
+
+/// A handle that signals the client connection to close.
+#[derive(Clone, Debug)]
+pub struct Close(Arc<Notify>);
+
+/// A handle to be notified when the client connection is complete.
+#[must_use = "Closed handle must be used"]
+#[derive(Debug)]
+pub struct Closed(Arc<Notify>);
+
+/// A middleware that adds a clone of the `ClientHandle` as an extension to each
+/// request.
+#[derive(Clone, Debug)]
+pub struct SetClientHandle<S> {
+    inner: S,
+    handle: ClientHandle,
+}
+
+// === ClientHandle ===
+
+impl AsRef<SocketAddr> for ClientHandle {
+    fn as_ref(&self) -> &SocketAddr {
+        &self.addr
+    }
+}
+
+impl Into<SocketAddr> for ClientHandle {
+    fn into(self) -> SocketAddr {
+        self.addr
+    }
+}
+
+// === Close ===
+
+impl Close {
+    pub fn close(&self) {
+        self.0.notify()
+    }
+}
+
+// === Closed ===
+
+impl Closed {
+    pub async fn closed(self) {
+        self.0.notified().await
+    }
+
+    pub fn ignore(self) {}
```

Comment:
Keeping this for future use?

Reply:
Yeah, since `Closed` has a `must_use`, this seemed like a nice convenience.